Make template gallery

mariaschuld opened the pull request:
**Context:**

The templates.rst file was outdated and needed the latest templates to be added. In the process I turned tables into a gallery, which is much more appealing. 

**Description of the Change:**

Add custom directives for gallery items and turn tables into gallery.

Results only affect this tab:

![Screenshot from 2020-02-06 12-15-21](https://user-images.githubusercontent.com/23052011/73927661-728af080-48da-11ea-919e-ee161c150082.png)

Here is a sneak peak:

![Screenshot from 2020-02-06 12-15-05](https://user-images.githubusercontent.com/23052011/73927630-643cd480-48da-11ea-8831-a9df517da029.png)
![Screenshot from 2020-02-06 12-15-08](https://user-images.githubusercontent.com/23052011/73927649-6e5ed300-48da-11ea-88be-3961de64d84d.png)
![Screenshot from 2020-02-06 12-15-16](https://user-images.githubusercontent.com/23052011/73927657-70c12d00-48da-11ea-99c0-4856f8a6a400.png)


**Benefits:** User gets an idea of what the templates are on first glance.

**Possible Drawbacks:** We should make the gallery thumbnails wider to have less white space? 

**Related GitHub Issues:** -

mariaschuld commented on doc/conf.py:
```diff
@@ -405,8 +405,9 @@ def __getattr__(cls, name):
 def setup(app):
+    app.add_directive('customgalleryitem', CustomGalleryItemDirective)
     app.add_directive("usagedetails", UsageDetails)
+    app.add_stylesheet('xanadu_gallery.css')
```
@josh146 not sure if that is needed?

josh146 replied:
I think it is - this is the file that determines the look and style of the cards.

josh146 replied:
Oh nice, I had forgotten I had already added the usage details role!

josh146 commented on doc/directives.py:
```diff
@@ -53,3 +56,166 @@ def run(self):
+                # if 'size' in self.options:
+                #     x, y = self.options['size'].split(" ")
+
+                # sphinx_gallery.gen_rst.scale_image(figname, thumbnail, int(x), int(y))
```
Was this commented out from the walrus? If so, I can't remember why I did this. I think to avoid the thumbnails from automatically being scaled?

josh146 replied:
We might as well delete it

josh146 commented on doc/directives.py:
```diff
@@ -53,3 +56,166 @@ def run(self):
+    option_spec = {'tooltip': directives.unchanged,
+                   'figure': directives.unchanged,
+                   'description': directives.unchanged,
+                   'size': directives.unchanged}
```
Ah, that's right, I made it so you could set the size explicitly I think...

josh146 replied:
Note: do we want to remove the tooltips? They seem a bit superfluous. If so, it should be sufficient to remove the `tooltip="{tooltip}"` from line 137, the tooltip option from lines 157, 168, and 215, as well as the if statement on lines 178-181

josh146 commented on doc/introduction/templates.rst:
```diff
@@ -74,8 +80,51 @@ functions that act on templates, such as :func:`~.pennylane.inv`:
+.. customgalleryitem::
+    :tooltip: AmplitudeEmbedding template
+    :description: :doc:`AmplitudeEmbedding</code/api/pennylane.templates.embeddings.AmplitudeEmbedding>`
+    :figure: ../_static/templates/embeddings/amplitude.png
```
All the thumbnails look great, but amplitude embedding is a bit vague (just a white square), is there a way of displaying slightly more info? Not a major comment though.

mariaschuld replied:
I thought it forth and back, and AmplitudeEmbedding is not defined by a circuit, and could in future actually depend on different execution methods depending on the device. So for all purposes it is here a black box...

johannesjmeyer commented on doc/introduction/templates.rst:
```diff
@@ -86,65 +135,64 @@ Layer templates
+------------------
 
-:html:`<div class="summary-table">`
+Subroutines are simply sequences of (possibly trainable) gates.
```
I think this is a copy-paste error from the description of the subroutines!

johannesjmeyer replied:
```suggestion
State preparation templates transform a given state into a sequence of gates preparing that state.
```

antalszava commented on doc/_static/xanadu_gallery.css:
```diff
@@ -211,55 +225,15 @@ p.sphx-glr-signature a.reference.external {
+  margin-top: 20px;
+  text-align: center;
+  font-size: small;
+}
```
```suggestion
}

```

mariaschuld commented:
> This is really beautiful, I love it!
> 
> Note, when I build the docs, I get the following errors:
> 
> ```shell
> /home/josh/Dropbox/Work/Xanadu/pennylane/doc/introduction/templates.rst:: WARNING: image file not readable: ../_static/templates/embeddings/amplitude.png
> /home/josh/Dropbox/Work/Xanadu/pennylane/doc/introduction/templates.rst:: WARNING: image file not readable: ../_static/templates/embeddings/angle.png
> /home/josh/Dropbox/Work/Xanadu/pennylane/doc/introduction/templates.rst:: WARNING: image file not readable: ../_static/templates/embeddings/basis.png
> /home/josh/Dropbox/Work/Xanadu/pennylane/doc/introduction/templates.rst:: WARNING: image file not readable: ../_static/templates/embeddings/displacement.png
> /home/josh/Dropbox/Work/Xanadu/pennylane/doc/introduction/templates.rst:: WARNING: image file not readable: ../_static/templates/embeddings/qaoa.png
> /home/josh/Dropbox/Work/Xanadu/pennylane/doc/introduction/templates.rst:: WARNING: image file not readable: ../_static/templates/embeddings/squeezing.png
> /home/josh/Dropbox/Work/Xanadu/pennylane/doc/introduction/templates.rst:: WARNING: image file not readable: ../_static/templates/layers/cvqnn.png
> /home/josh/Dropbox/Work/Xanadu/pennylane/doc/introduction/templates.rst:: WARNING: image file not readable: ../_static/templates/layers/random.png
> /home/josh/Dropbox/Work/Xanadu/pennylane/doc/introduction/templates.rst:: WARNING: image file not readable: ../_static/templates/layers/strongly_entangling.png
> /home/josh/Dropbox/Work/Xanadu/pennylane/doc/introduction/templates.rst:: WARNING: image file not readable: ../_static/templates/state_preparations/basis.png
> /home/josh/Dropbox/Work/Xanadu/pennylane/doc/introduction/templates.rst:: WARNING: image file not readable: ../_static/templates/state_preparations/mottonen.png
> /home/josh/Dropbox/Work/Xanadu/pennylane/doc/introduction/templates.rst:: WARNING: image file not readable: ../_static/templates/subroutines/interferometer.png
> ```
> 
> An issue with the path? Note that if you instead do `/_static`, sphinx treats this as an absolute path starting from the doc folder (avoiding relative path headaches).
> 
> However, I notice now that I can't see these 'failed' images anywhere, everything seems to be working...

I tried for hours and cannot figure out why sphinx is behaving this way. Other options for the path are also not working. Maybe related to this: https://github.com/sphinx-gallery/sphinx-gallery/issues/120 ?

josh146 commented:
I'll try and look into it today!

mariaschuld commented:
Runs without errors now, I used some code that was commented out in the walrus - I think sphinx takes images and saves them again in a reshaped/thumb version, which is the one it wants to import.

mariaschuld commented:
From my side the only item on my wishlist is to make the images in the gallery items larger. I managed to rescale the cards themselves to be wider, but I cannot enlarge the image.
This is what I tired:

* use sphinx_gallery.gen_rst.scale_image -> nothing happens
* use :width: in html template -> shifts and rescales in a very strange fashion
* change css -> only affects container

mariaschuld commented:
Switched to cards and updated some pictures...

Looking much better now!

![Screenshot from 2020-02-10 23-03-00](https://user-images.githubusercontent.com/23052011/74189756-a6a73e00-4c59-11ea-8cc8-fd2f8eb53002.png)

josh146 commented:
Awesome! Couple of quick questions:

* How does it deal with responsiveness? I.e. as you make the browser width smaller and smaller, does it scale properly?

* Some of the cards seem taller than others?

mariaschuld commented:
Scaling works ok, the cards move into double and then single columns. The is a glitch though, it doesn't fill up the last row nicely:

![Screenshot from 2020-02-11 16-48-04](https://user-images.githubusercontent.com/23052011/74247183-5b8b3a80-4cee-11ea-8b5b-ea389b6009ba.png)

Single column works well

![Screenshot from 2020-02-11 16-49-09](https://user-images.githubusercontent.com/23052011/74247255-76f64580-4cee-11ea-9989-30329e2f28e1.png)

As for the size, I can't see a difference between cards. Which one did you spot? Should be regulated by the picture size, which should be the same by design.

josh146 commented:
The three layers seem very subtly different heights (e.g., the baseline of StrongEnt seems to be below that of RandomLayers). A height difference could also explain that missing hole when resizing the template cards

mariaschuld commented:
You were right @josh146. With the height and width of the image fixed the resizing problem disappears.

![Uploading Screenshot from 2020-02-14 08-52-07.png…]()

And this is the normal view:

![Screenshot from 2020-02-14 08-52-24](https://user-images.githubusercontent.com/23052011/74508328-797fb780-4f07-11ea-969b-19fe0e998ab7.png)

josh146 commented on doc/directives.py:
```diff
@@ -57,9 +57,9 @@ def run(self):
+    <div class="card" style="width: 13rem; float:left; margin: 10px;">
         <a href={link}>
-            <img class="card-img-top" src={thumbnail} alt="image not found">
+            <img class="card-img-top" src={thumbnail} alt="image not found" style="width: 13rem; height: 6rem;">
```
:+1:
